### PR TITLE
router: preserve retry flags when retriable req headers are used 

### PR DESCRIPTION
**Pull request by snowp**
Previously, when a retriable request header was found the retry bitset
would be AND'd with 1, which meant that only the 0x1 flag (5xx) would be
retained. This PR updates it to instead clear out the bitset when no
flags are missing - this should ensure that all retry flags are
preserved when using retriable request headers.

Also removes retriable_request_headers from the RetryPolicyImpl type -
they were only being read in the ctor.

Signed-off-by: Snow Pettersen <aickck@gmail.com>

Risk Level: Low
Testing: UT
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: test/common/router/retry_state_impl_test.cc [hunk 1]
```diff
@@ -704,16 +704,32 @@ TEST_F(RouterRetryStateImplTest, PolicyLimitedByRequestHeaders) {
...
+    EXPECT_EQ(RetryStatus::Yes, state_->shouldRetryHeaders(response_headers, callback_));
   }
 
   {
     Http::TestHeaderMapImpl request_headers{{":method", "HEAD"}, {"x-envoy-retry-on", "5xx"}};
     setup(request_headers);
     EXPECT_TRUE(state_->enabled());
+    Http::TestHeaderMapImpl response_headers{{":status", "500"}};
+    expectTimerCreateAndEnable();
+    EXPECT_EQ(RetryStatus::Yes, state_->shouldRetryHeaders(response_headers, callback_));
```
Comment (junr03) [inline test/common/router/retry_state_impl_test.cc]: Should we also have a test where the policy is enabled but the RetryStatus is No?

Comment (snowp) [reply test/common/router/retry_state_impl_test.cc]: sure ill add one

**Comment by alyssawilk** [comment]
Mind checking out the test failures?  I'm happy to approve once CI is happy :-)

**Comment by alyssawilk** [comment]
Oh yeah, and still not happy.  Alas!

**Comment by snowp** [comment]
Hopefully should be good now! Sorry for the repeated failed builds :)

**Pull request merged**